File: Cuarto/Programación Distribuida/python_concurrencia2/ejercicio1.py
import threading
import random
import time

class Productor(threading.Thread):
    def __init__(self, buffer):
        super().__init__()
        self.buffer = buffer

    def run(self):
        while True:
            self.buffer.add()
            time.sleep(1)

class Consumidor(threading.Thread):
    def __init__(self, buffer, id):
        super().__init__()
        self.buffer = buffer
        self.id = id

    def run(self):
        while True:
            elem = self.buffer.remove()
            print(elem, " from ", self.id)
            time.sleep(2)

class Buffer():
    def __init__(self):
        self.buffer = []
        self.cond = threading.Condition()

    def add(self):
        self.cond.acquire()
        if len(self.buffer) == 10: self.cond.wait()

        self.buffer.append(random.randint(0,100))

        if (len(self.buffer) == 1): self.cond.notify()
        self.cond.release()
    
    def remove(self):
        self.cond.acquire()
        if len(self.buffer) == 0: self.cond.wait()

        elem = self.buffer.pop()

        if (len(self.buffer) == 9): self.cond.notify()
        self.cond.release()

        return elem

buffer = Buffer()
thread_productor = Productor(buffer)
thread_productor.start()

thread_consumidor1 = Consumidor(buffer, 1)
thread_consumidor2 = Consumidor(buffer, 2)

thread_consumidor1.start()
thread_consumidor2.start()

File: Cuarto/ProgramaciónDistribuida/java_concurrencia2/ejercicio1.java
package java_concurrencia2;

import java.util.LinkedList;
import java.util.List;
import java.util.Random;

public class ejercicio1 {
    static List<Integer> list;

    public static class Buffer{
        private List<Integer> list;
        public Buffer(){
            list = new LinkedList<>();
        }

        public synchronized void add(){
            if (list.size() == 10) {
                try {
                    wait();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
            Random rand = new Random();
            list.add(rand.nextInt(100));
            if (list.size() == 1) notify();
        }

        public synchronized int remove(){
            if (list.size() == 0){ 
                try {
                    wait();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                } 
            }
            int elem = list.remove(list.size() - 1);
            if (list.size() == 9) notify();
            return elem;
        }
    }

    public static class Productor extends Thread {
        public Productor() {}

        public void run(){
            while(true){
                buffer.add();
                try {
                    sleep(1000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    public static class Consumidor extends Thread {
        private int id;
        public Consumidor(int id) { this.id = id;}

        public void run(){
            while(true){
                int elem = buffer.remove();
                System.out.println(elem + " from " + id);
                try {
                    sleep(2000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    static Buffer buffer = new Buffer();

    public static void main(String[] args) {
        Productor p = new Productor();
        list = new LinkedList<>();
        p.start();
        Consumidor c1 = new Consumidor(1);
        Consumidor c2 = new Consumidor(2);
        
        c1.start();
        c2.start();
    }
}
